pythonopensubtitles/opensubtitles.py

```diff
@@ -104,9 +109,99 @@ def detect_language(self):
+
+        subfile_ids can be None/empty if a filenames dict is present;
+        otherwise, the file IDs used by OpenSubtitles are used as file
+        names, with .srt as default file extension.
+
+        Keep in mind that subtitle searches can contain multiple
+        results for the same file title (because several different
+        file uploads might reference the same file name).
+        So, if you use a filenames dict in which different file IDs
+        reference the exact same file name (i.e. IDs with the same value),
+        the number of eventually downloaded files will be smaller than the
+        overall number of file IDs submitted (there can only be one file
+        per file name, so any subsequent download attempts for an already
+        existing file will overwrite the previous version of said file).
+
+        :param subfile_ids: list of ids of subtitle files to download
+        :param filenames: optional dictionary with preferred file names;
+                          for keys use subtitle file IDs (as strings),
+                          for values use file names (incl. file extensions)
+        :param ext: one of: srt, sub, txt, ssa, smi, mpi, tmp;
+                    only applicable if only subtitle file IDs are provided
+        :param directory: path to preferred dir to which to write files
+        """
+        successful = {}
+        if filenames:
+            subfile_ids = list(filenames.keys())
+
+        # OpenSubtitles will accept a maximum of 20 IDs for download
+        if len(subfile_ids) > 20:
+            print("Cannot download more than 20 files at once.",
+                  file=sys.stderr)
+            subfile_ids = subfile_ids[:20]
+
+        self.data = self.xmlrpc.DownloadSubtitles(self.token, subfile_ids)
+        encoded_data = self._get_from_data_or_none('data')
+
+        if encoded_data:
+            for item in encoded_data:
+                subfile_id = item['idsubtitlefile']
+
+                try:
+                    decoded_data = zlib.decompress(
+                        base64.b64decode(item['data']),
+                        16 + zlib.MAX_WBITS).decode('utf-8')
+                except UnicodeDecodeError as e:
+                    try:
+                        decoded_data = zlib.decompress(
+                            base64.b64decode(item['data']),
+                            16 + zlib.MAX_WBITS).decode('latin1')
+                    except Exception as e:
+                        print("An error occurred during file decoding.",
+                              file=sys.stderr)
+                        print(e)
+                        continue
+                except Exception as e:
+                    print("An error occurred during file decoding.",
+                          file=sys.stderr)
+                    print(e)
+                    continue
+
+                if filenames:
+                    try:
+                        fpath = filenames[subfile_id]
```

Comment:
If you remember at the beginning I asked you to init filenames as an empty dict in case it was send as `None`, that would make this piece of code a little bit easier:

```
fpath = filenames.get(subfile_id, subfile_id + '.' + ext)
```

Basically it's trying to get `subfiled_id` key from the dict and if it isn't present it will default to `subfile_id + '.' + ext`.

Reply:
Missed one: this is nice! I didn't know about `get`, or if I did at some point in the past, I forgot about it again. (I've used `getenv` though, which seems really similar/which this reminds me of.)